common/src/main/java/io/druid/collections/BlockingPool.java

```diff
@@ -74,40 +74,56 @@ public int getPoolSize()
   }
 
   /**
-   * Take a resource from the pool.
+   * Take a resource from the pool, waiting up to the
+   * specified wait time if necessary for an element to become available.
    *
-   * @param timeout maximum time to wait for a resource, in milliseconds. Negative means do not use a timeout.
+   * @param timeout maximum time to wait for a resource, in milliseconds.
    *
    * @return a resource, or null if the timeout was reached
    */
   public ReferenceCountingResourceHolder<T> take(final long timeout)
   {
     checkInitialized();
-    final T theObject;
     try {
-      if (timeout > -1) {
-        theObject = timeout > 0 ? poll(timeout) : poll();
-      } else {
-        theObject = take();
-      }
-      return theObject == null ? null : new ReferenceCountingResourceHolder<>(
-          theObject,
-          new Closeable()
-          {
-            @Override
-            public void close() throws IOException
-            {
-              offer(theObject);
-            }
-          }
-      );
+      return wrapObject(timeout > 0 ? pollObject(timeout) : pollObject());
```

Comment:
Since there is a dedicated no-timeout method, in this method I suggest to check if timeout is positive and throw an exception otherwise.